File: plugin/Functions/AdvancedGeometry.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;
using Rhino;
using Rhino.DocObjects;
using Rhino.Geometry;

namespace RhinoMCPPlugin.Functions;

/// <summary>
/// Advanced geometry operations: Loft, Extrude, Sweep, Offset, Pipe
/// </summary>
public partial class RhinoMCPFunctions
{
    /// <summary>
    /// Create a loft surface through multiple curves.
    /// </summary>
    [McpCommand("loft")]
    public JObject Loft(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        var curveIds = parameters["curve_ids"]?.ToObject<List<string>>();
        var name = parameters["name"]?.ToString();
        var closed = parameters["closed"]?.ToObject<bool>() ?? false;
        var loftType = parameters["loft_type"]?.ToObject<int>() ?? 0;

        if (curveIds == null || curveIds.Count < 2)
            throw new ArgumentException("Loft requires at least 2 curve IDs");

        var curves = new List<Curve>();
        foreach (var idStr in curveIds)
        {
            var obj = doc.Objects.Find(new Guid(idStr));
            if (obj == null)
                throw new InvalidOperationException($"Object with ID {idStr} not found");

            var curve = obj.Geometry as Curve;
            if (curve == null)
                throw new InvalidOperationException($"Object {idStr} is not a curve");

            curves.Add(curve);
        }

        // Convert loft type
        var loftTypeEnum = loftType switch
        {
            0 => LoftType.Normal,
            1 => LoftType.Loose,
            2 => LoftType.Tight,
            3 => LoftType.Straight,
            _ => LoftType.Normal
        };

        var breps = Brep.CreateFromLoft(curves, Point3d.Unset, Point3d.Unset, loftTypeEnum, closed);

        if (breps == null || breps.Length == 0)
            throw new InvalidOperationException("Loft operation failed - ensure curves are valid and compatible");

        var resultIds = new JArray();
        foreach (var brep in breps)
        {
            var attr = new ObjectAttributes();
            if (!string.IsNullOrEmpty(name))
                attr.Name = name;

            var id = doc.Objects.AddBrep(brep, attr);
            resultIds.Add(id.ToString());
        }

        doc.Views.Redraw();

        return new JObject
        {
            ["result_ids"] = resultIds,
            ["count"] = breps.Length,
            ["message"] = $"Loft created {breps.Length} surface(s)"
        };
    }

    /// <summary>
    /// Extrude a curve along a direction vector.
    /// </summary>
    [McpCommand("extrude_curve")]
    public JObject ExtrudeCurve(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        var curveId = parameters["curve_id"]?.ToString();
        var direction = parameters["direction"]?.ToObject<List<double>>();
        var name = parameters["name"]?.ToString();
        var cap = parameters["cap"]?.ToObject<bool>() ?? true;

        if (string.IsNullOrEmpty(curveId))
            throw new ArgumentException("Extrude requires a curve_id");

        if (direction == null || direction.Count != 3)
            throw new ArgumentException("Extrude requires a direction vector [x, y, z]");

        var obj = doc.Objects.Find(new Guid(curveId));
        if (obj == null)
            throw new InvalidOperationException($"Object with ID {curveId} not found");

        var curve = obj.Geometry as Curve;
        if (curve == null)
            throw new InvalidOperationException($"Object {curveId} is not a curve");

        var directionVec = new Vector3d(direction[0], direction[1], direction[2]);

        // Create extrusion
        Surface extrudedSurface = Surface.CreateExtrusion(curve, directionVec);
        if (extrudedSurface == null)
            throw new InvalidOperationException("Extrusion failed");

        Guid resultId;
        var attr = new ObjectAttributes();
        if (!string.IsNullOrEmpty(name))
            attr.Name = name;

        // If curve is closed and cap is requested, create a capped solid
        if (cap && curve.IsClosed)
        {
            var brep = extrudedSurface.ToBrep();
            if (brep != null)
            {
                var cappedBrep = brep.CapPlanarHoles(doc.ModelAbsoluteTolerance);
                if (cappedBrep != null)
                {
                    resultId = doc.Objects.AddBrep(cappedBrep, attr);
                }
                else
                {
                    resultId = doc.Objects.AddBrep(brep, attr);
                }
            }
            else
            {
                resultId = doc.Objects.AddSurface(extrudedSurface, attr);
            }
        }
        else
        {
            resultId = doc.Objects.AddSurface(extrudedSurface, attr);
        }

        doc.Views.Redraw();

        return new JObject
        {
            ["result_id"] = resultId.ToString(),
            ["message"] = "Extrusion created successfully"
        };
    }

    /// <summary>
    /// Sweep profile curves along a rail curve.
    /// </summary>
    [McpCommand("sweep1")]
    public JObject Sweep1(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        var railId = parameters["rail_id"]?.ToString();
        var profileIds = parameters["profile_ids"]?.ToObject<List<string>>();
        var name = parameters["name"]?.ToString();
        var closed = parameters["closed"]?.ToObject<bool>() ?? false;

        if (string.IsNullOrEmpty(railId))
            throw new ArgumentException("Sweep requires a rail_id");

        if (profileIds == null || profileIds.Count == 0)
            throw new ArgumentException("Sweep requires at least one profile curve");

        // Get rail curve
        var railObj = doc.Objects.Find(new Guid(railId));
        if (railObj == null)
            throw new InvalidOperationException($"Rail object with ID {railId} not found");

        var rail = railObj.Geometry as Curve;
        if (rail == null)
            throw new InvalidOperationException($"Rail object {railId} is not a curve");

        // Get profile curves
        var profiles = new List<Curve>();
        foreach (var idStr in profileIds)
        {
            var obj = doc.Objects.Find(new Guid(idStr));
            if (obj == null)
                throw new InvalidOperationException($"Profile object with ID {idStr} not found");

            var curve = obj.Geometry as Curve;
            if (curve == null)
                throw new InvalidOperationException($"Profile object {idStr} is not a curve");

            profiles.Add(curve);
        }

        // Create sweep
        var sweep = new SweepOneRail();
        sweep.SetToRoadlikeTop();
        var breps = sweep.PerformSweep(rail, profiles);

        if (breps == null || breps.Length == 0)
            throw new InvalidOperationException("Sweep operation failed - ensure rail and profiles are valid");

        var resultIds = new JArray();
        foreach (var brep in breps)
        {
            var attr = new ObjectAttributes();
            if (!string.IsNullOrEmpty(name))
                attr.Name = name;

            var id = doc.Objects.AddBrep(brep, attr);
            resultIds.Add(id.ToString());
        }

        doc.Views.Redraw();

        return new JObject
        {
            ["result_ids"] = resultIds,
            ["count"] = breps.Length,
            ["message"] = $"Sweep created {breps.Length} surface(s)"
        };
    }

    /// <summary>
    /// Offset a curve by a specified distance.
    /// </summary>
    [McpCommand("offset_curve")]
    public JObject OffsetCurve(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        var curveId = parameters["curve_id"]?.ToString();
        var distance = parameters["distance"]?.ToObject<double>() ?? 0;
        var name = parameters["name"]?.ToString();
        var planeNormal = parameters["plane"]?.ToObject<List<double>>();
        var cornerStyle = parameters["corner_style"]?.ToObject<int>() ?? 1;

        if (string.IsNullOrEmpty(curveId))
            throw new ArgumentException("Offset requires a curve_id");

        if (Math.Abs(distance) < doc.ModelAbsoluteTolerance)
            throw new ArgumentException("Offset distance is too small");

        var obj = doc.Objects.Find(new Guid(curveId));
        if (obj == null)
            throw new InvalidOperationException($"Object with ID {curveId} not found");

        var curve = obj.Geometry as Curve;
        if (curve == null)
            throw new InvalidOperationException($"Object {curveId} is not a curve");

        // Determine the offset plane
        Plane offsetPlane;
        if (planeNormal != null && planeNormal.Count == 3)
        {
            var normal = new Vector3d(planeNormal[0], planeNormal[1], planeNormal[2]);
            offsetPlane = new Plane(curve.PointAtStart, normal);
        }
        else
        {
            // Try to get the curve's plane, fall back to world XY
            if (!curve.TryGetPlane(out offsetPlane))
            {
                offsetPlane = Plane.WorldXY;
                offsetPlane.Origin = curve.PointAtStart;
            }
        }

        // Convert corner style
        var cornerStyleEnum = cornerStyle switch
        {
            0 => CurveOffsetCornerStyle.None,
            1 => CurveOffsetCornerStyle.Sharp,
            2 => CurveOffsetCornerStyle.Round,
            3 => CurveOffsetCornerStyle.Smooth,
            4 => CurveOffsetCornerStyle.Chamfer,
            _ => CurveOffsetCornerStyle.Sharp
        };

        var offsetCurves = curve.Offset(offsetPlane, distance, doc.ModelAbsoluteTolerance, cornerStyleEnum);

        if (offsetCurves == null || offsetCurves.Length == 0)
            throw new InvalidOperationException("Offset operation failed");

        var resultIds = new JArray();
        foreach (var offsetCurve in offsetCurves)
        {
            var attr = new ObjectAttributes();
            if (!string.IsNullOrEmpty(name))
                attr.Name = name;

            var id = doc.Objects.AddCurve(offsetCurve, attr);
            resultIds.Add(id.ToString());
        }

        doc.Views.Redraw();

        return new JObject
        {
            ["result_ids"] = resultIds,
            ["count"] = offsetCurves.Length,
            ["message"] = $"Offset created {offsetCurves.Length} curve(s)"
        };
    }

    /// <summary>
    /// Create a pipe along a curve.
    /// </summary>
    [McpCommand("pipe")]
    public JObject Pipe(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        var curveId = parameters["curve_id"]?.ToString();
        var radius = parameters["radius"]?.ToObject<double>() ?? 0;
        var name = parameters["name"]?.ToString();
        var cap = parameters["cap"]?.ToObject<bool>() ?? true;
        var fitRail = parameters["fit_rail"]?.ToObject<bool>() ?? false;

        if (string.IsNullOrEmpty(curveId))
            throw new ArgumentException("Pipe requires a curve_id");

        if (radius <= 0)
            throw new ArgumentException("Pipe radius must be positive");

        var obj = doc.Objects.Find(new Guid(curveId));
        if (obj == null)
            throw new InvalidOperationException($"Object with ID {curveId} not found");

        var curve = obj.Geometry as Curve;
        if (curve == null)
            throw new InvalidOperationException($"Object {curveId} is not a curve");

        // Determine cap type
        var capType = cap ? PipeCapMode.Flat : PipeCapMode.None;

        // Create pipe with constant radius
        var breps = Brep.CreatePipe(
            curve,
            radius,
            localBlending: !fitRail,
            cap: capType,
            fitRail: fitRail,
            absoluteTolerance: doc.ModelAbsoluteTolerance,
            angleToleranceRadians: doc.ModelAngleToleranceRadians
        );

        if (breps == null || breps.Length == 0)
            throw new InvalidOperationException("Pipe operation failed");

        var resultIds = new JArray();
        foreach (var brep in breps)
        {
            var attr = new ObjectAttributes();
            if (!string.IsNullOrEmpty(name))
                attr.Name = name;

            var id = doc.Objects.AddBrep(brep, attr);
            resultIds.Add(id.ToString());
        }

        doc.Views.Redraw();

        return new JObject
        {
            ["result_ids"] = resultIds,
            ["count"] = breps.Length,
            ["message"] = $"Pipe created {breps.Length} object(s)"
        };
    }
}

File: server/src/rhinomcp/tools/advanced_geometry.py
"""
Advanced geometry tools for common operations that would otherwise require RhinoScript.

These tools provide a simple, hallucination-free interface for:
- Loft: Create surface from multiple curves
- Extrude: Extend a curve along a direction
- Sweep1: Sweep profiles along a rail curve
- Offset: Offset a curve by a distance
- Pipe: Create a pipe along a curve
"""

from mcp.server.fastmcp import Context
from rhinomcp.server import get_rhino_connection, mcp, logger
from typing import List, Optional, Dict, Any


@mcp.tool()
def loft(
    ctx: Context,
    curve_ids: List[str],
    name: Optional[str] = None,
    closed: bool = False,
    loft_type: int = 0,
) -> Dict[str, Any]:
    """
    Create a loft surface through multiple curves.

    Parameters:
    - curve_ids: List of curve IDs (GUIDs) to loft through (minimum 2 curves).
                 Curves should be in order from start to end of the loft.
    - name: Optional name for the resulting surface
    - closed: If True, creates a closed loft (connects last curve back to first)
    - loft_type: Type of loft (0=Normal, 1=Loose, 2=Tight, 3=Straight, 4=Developable)

    Returns:
    - Dictionary with result_ids (list of created surface IDs) and message

    Example:
        # Create two circles and loft between them
        circle1 = create_object(type="CIRCLE", params={"center": [0,0,0], "radius": 5})
        circle2 = create_object(type="CIRCLE", params={"center": [0,0,10], "radius": 3})
        result = loft(curve_ids=[circle1_id, circle2_id], name="my_loft")
    """
    try:
        if len(curve_ids) < 2:
            return {"success": False, "message": "Loft requires at least 2 curves"}

        rhino = get_rhino_connection()
        params = {
            "curve_ids": curve_ids,
            "closed": closed,
            "loft_type": loft_type,
        }
        if name:
            params["name"] = name

        result = rhino.send_command("loft", params)
        return {
            "success": True,
            "result_ids": result.get("result_ids", []),
            "message": result.get("message", "Loft created successfully")
        }
    except Exception as e:
        logger.error(f"Error in loft: {str(e)}")
        return {"success": False, "message": str(e)}


@mcp.tool()
def extrude_curve(
    ctx: Context,
    curve_id: str,
    direction: List[float],
    name: Optional[str] = None,
    cap: bool = True,
) -> Dict[str, Any]:
    """
    Extrude a curve along a direction vector to create a surface or solid.

    Parameters:
    - curve_id: The curve ID (GUID) to extrude
    - direction: Extrusion direction as [x, y, z] vector (e.g., [0, 0, 10] for 10 units up)
    - name: Optional name for the resulting object
    - cap: If True and curve is closed, caps the ends to create a solid (default: True)

    Returns:
    - Dictionary with result_id (created object ID) and message

    Example:
        # Create a circle and extrude it upward to make a cylinder
        circle = create_object(type="CIRCLE", params={"center": [0,0,0], "radius": 5})
        result = extrude_curve(curve_id=circle_id, direction=[0, 0, 10], name="cylinder")
    """
    try:
        if len(direction) != 3:
            return {"success": False, "message": "Direction must be [x, y, z] vector"}

        rhino = get_rhino_connection()
        params = {
            "curve_id": curve_id,
            "direction": direction,
            "cap": cap,
        }
        if name:
            params["name"] = name

        result = rhino.send_command("extrude_curve", params)
        return {
            "success": True,
            "result_id": result.get("result_id"),
            "message": result.get("message", "Extrusion created successfully")
        }
    except Exception as e:
        logger.error(f"Error in extrude_curve: {str(e)}")
        return {"success": False, "message": str(e)}


@mcp.tool()
def sweep1(
    ctx: Context,
    rail_id: str,
    profile_ids: List[str],
    name: Optional[str] = None,
    closed: bool = False,
) -> Dict[str, Any]:
    """
    Sweep one or more profile curves along a rail curve.

    Parameters:
    - rail_id: The rail curve ID (GUID) - the path along which to sweep
    - profile_ids: List of profile curve IDs (GUIDs) - the shapes to sweep
    - name: Optional name for the resulting surface
    - closed: If True, creates a closed sweep

    Returns:
    - Dictionary with result_ids (list of created surface IDs) and message

    Example:
        # Create a rail (line) and profile (circle), then sweep
        rail = create_object(type="LINE", params={"start": [0,0,0], "end": [10,0,5]})
        profile = create_object(type="CIRCLE", params={"center": [0,0,0], "radius": 1})
        result = sweep1(rail_id=rail_id, profile_ids=[profile_id], name="swept_tube")
    """
    try:
        if not profile_ids:
            return {"success": False, "message": "Sweep requires at least 1 profile curve"}

        rhino = get_rhino_connection()
        params = {
            "rail_id": rail_id,
            "profile_ids": profile_ids,
            "closed": closed,
        }
        if name:
            params["name"] = name

        result = rhino.send_command("sweep1", params)
        return {
            "success": True,
            "result_ids": result.get("result_ids", []),
            "message": result.get("message", "Sweep created successfully")
        }
    except Exception as e:
        logger.error(f"Error in sweep1: {str(e)}")
        return {"success": False, "message": str(e)}


@mcp.tool()
def offset_curve(
    ctx: Context,
    curve_id: str,
    distance: float,
    name: Optional[str] = None,
    plane: Optional[List[float]] = None,
    corner_style: int = 1,
) -> Dict[str, Any]:
    """
    Offset a curve by a specified distance.

    Parameters:
    - curve_id: The curve ID (GUID) to offset
    - distance: Offset distance (positive = one side, negative = other side)
    - name: Optional name for the resulting curve
    - plane: Optional plane normal [x, y, z] for the offset direction.
             If not provided, uses the curve's plane or world XY.
    - corner_style: How to handle corners (0=None, 1=Sharp, 2=Round, 3=Smooth, 4=Chamfer)

    Returns:
    - Dictionary with result_ids (list of created curve IDs) and message

    Example:
        # Offset a curve by 2 units
        result = offset_curve(curve_id=my_curve_id, distance=2.0, name="offset_curve")
    """
    try:
        rhino = get_rhino_connection()
        params = {
            "curve_id": curve_id,
            "distance": distance,
            "corner_style": corner_style,
        }
        if name:
            params["name"] = name
        if plane:
            params["plane"] = plane

        result = rhino.send_command("offset_curve", params)
        return {
            "success": True,
            "result_ids": result.get("result_ids", []),
            "message": result.get("message", "Offset created successfully")
        }
    except Exception as e:
        logger.error(f"Error in offset_curve: {str(e)}")
        return {"success": False, "message": str(e)}


@mcp.tool()
def pipe(
    ctx: Context,
    curve_id: str,
    radius: float,
    name: Optional[str] = None,
    cap: bool = True,
    fit_rail: bool = False,
) -> Dict[str, Any]:
    """
    Create a pipe (tube) along a curve.

    Parameters:
    - curve_id: The curve ID (GUID) to create pipe along
    - radius: Radius of the pipe
    - name: Optional name for the resulting object
    - cap: If True, caps the ends of the pipe (default: True)
    - fit_rail: If True, fits the pipe more closely to the rail curve

    Returns:
    - Dictionary with result_ids (list of created object IDs) and message

    Example:
        # Create a curved line and make it into a pipe
        curve = create_object(type="ARC", params={"center": [0,0,0], "radius": 10, "start_angle": 0, "end_angle": 90})
        result = pipe(curve_id=curve_id, radius=0.5, name="pipe_tube")
    """
    try:
        if radius <= 0:
            return {"success": False, "message": "Radius must be positive"}

        rhino = get_rhino_connection()
        params = {
            "curve_id": curve_id,
            "radius": radius,
            "cap": cap,
            "fit_rail": fit_rail,
        }
        if name:
            params["name"] = name

        result = rhino.send_command("pipe", params)
        return {
            "success": True,
            "result_ids": result.get("result_ids", []),
            "message": result.get("message", "Pipe created successfully")
        }
    except Exception as e:
        logger.error(f"Error in pipe: {str(e)}")
        return {"success": False, "message": str(e)}
